File: src/remi/scheduler/jobs.py
"""Scheduler for periodic tasks using pure asyncio.

Jobs:
- Heartbeat: check connector/provider health
- Memory compaction: archive daily notes → long-term memory + entity extraction
- Cleanup: remove old dailies and version files
"""

from __future__ import annotations

import asyncio
import logging
import re
from datetime import datetime, timedelta
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from remi.config import RemiConfig
    from remi.core import Remi

logger = logging.getLogger(__name__)


def _parse_cron_hour(cron_expr: str) -> int:
    """Extract hour from simple cron expression like '0 3 * * *'."""
    parts = cron_expr.split()
    if len(parts) >= 2:
        try:
            return int(parts[1])
        except ValueError:
            pass
    return 3  # default: 3 AM


class Scheduler:
    """Simple asyncio-based scheduler for periodic tasks."""

    def __init__(self, remi: Remi, config: RemiConfig) -> None:
        self._remi = remi
        self._config = config
        self._compact_hour = _parse_cron_hour(config.scheduler.memory_compact_cron)
        self._heartbeat_interval = config.scheduler.heartbeat_interval

    async def start(self, shutdown_event: asyncio.Event) -> None:
        """Run scheduled jobs until shutdown_event is set."""
        logger.info(
            "Scheduler started (heartbeat=%ds, compact@%02d:00)",
            self._heartbeat_interval,
            self._compact_hour,
        )

        last_compact_date: str | None = None

        while not shutdown_event.is_set():
            try:
                await asyncio.wait_for(
                    shutdown_event.wait(),
                    timeout=self._heartbeat_interval,
                )
                break  # shutdown requested
            except asyncio.TimeoutError:
                pass  # interval elapsed — run jobs

            # Heartbeat
            await self._heartbeat()

            # Daily compaction (run once per day at configured hour)
            now = datetime.now()
            today = now.strftime("%Y-%m-%d")
            if now.hour == self._compact_hour and last_compact_date != today:
                await self._compact_memory()
                await self._cleanup()
                last_compact_date = today

        logger.info("Scheduler stopped.")

    async def _heartbeat(self) -> None:
        """Check provider health."""
        for name, provider in self._remi._providers.items():
            try:
                healthy = await provider.health_check()
                if not healthy:
                    logger.warning("Provider %s health check failed", name)
            except Exception as e:
                logger.error("Provider %s health check error: %s", name, e)

    async def _compact_memory(self) -> None:
        """Summarize yesterday's daily notes, extract entities, update rolling summary.

        v2 enhanced compaction:
        1. Read yesterday's daily log
        2. Append observations to mentioned entities
        3. Extract new entities
        4. Update rolling summary (.conversation_summary.md)
        5. Compress 8-30 day logs into weekly summaries
        6. Archive logs older than 30 days
        """
        yesterday = (datetime.now() - timedelta(days=1)).strftime("%Y-%m-%d")
        daily = self._remi.memory.read_daily(yesterday)

        if not daily or len(daily.strip()) < 50:
            logger.debug("No significant daily notes to compact for %s", yesterday)
            return

        logger.info("Compacting daily notes for %s", yesterday)

        try:
            provider = self._remi._get_provider()
            prompt = (
                f"Below are my daily notes from {yesterday}. "
                "Extract any important facts, decisions, or preferences that should be "
                "remembered long-term. Format as bullet points. "
                "If nothing is worth remembering long-term, respond with 'SKIP'.\n\n"
                f"Also identify any people, organizations, or decisions mentioned. "
                f"For each, output a line: ENTITY: name (type) - observation\n\n"
                f"{daily}"
            )
            response = await provider.send(prompt)
            text = response.text.strip()

            if text.upper() != "SKIP":
                # Extract entity observations
                for line in text.splitlines():
                    if line.startswith("ENTITY:"):
                        self._process_entity_line(line)

                # Filter out ENTITY lines for the general summary
                summary_lines = [
                    line for line in text.splitlines() if not line.startswith("ENTITY:")
                ]
                summary_text = "\n".join(summary_lines).strip()

                if summary_text:
                    self._remi.memory.append_memory(f"\n## From {yesterday}\n\n{summary_text}")
                    logger.info("Appended compacted memory from %s", yesterday)

                # Update rolling summary
                self._update_rolling_summary(yesterday, summary_text)

            # Compress old daily logs into weekly summaries
            self._compress_weekly_logs()

            # Archive very old logs
            self._archive_old_logs()

        except Exception as e:
            logger.error("Memory compaction failed: %s", e)

    def _process_entity_line(self, line: str) -> None:
        """Parse and process an ENTITY: line from compaction output."""
        # Format: ENTITY: name (type) - observation
        match = re.match(r"ENTITY:\s*(.+?)\s*\((\w+)\)\s*-\s*(.+)", line)
        if not match:
            return
        name, etype, observation = match.group(1), match.group(2), match.group(3)
        try:
            entity_path = self._remi.memory._find_entity_by_name(name)
            if entity_path:
                self._remi.memory.append_observation(name, observation)
            else:
                self._remi.memory.create_entity(
                    name=name,
                    type=etype,
                    content=observation,
                    source="agent-inferred",
                )
        except Exception as e:
            logger.warning("Failed to process entity %s: %s", name, e)

    def _update_rolling_summary(self, date_str: str, summary: str) -> None:
        """Update the rolling conversation summary file."""
        summary_file = self._remi.memory.root / ".conversation_summary.md"
        try:
            existing = ""
            if summary_file.exists():
                existing = summary_file.read_text(encoding="utf-8")

            entry = f"\n## {date_str}\n{summary}\n"
            summary_file.write_text(existing + entry, encoding="utf-8")
        except OSError as e:
            logger.warning("Failed to update rolling summary: %s", e)

    def _compress_weekly_logs(self) -> None:
        """Compress 8-30 day old daily logs into weekly summaries."""
        daily_dir = self._remi.memory.root / "daily"
        now = datetime.now()

        for path in sorted(daily_dir.glob("*.md")):
            if path.stem.startswith("weekly-"):
                continue
            try:
                log_date = datetime.strptime(path.stem, "%Y-%m-%d")
            except ValueError:
                continue

            age = (now - log_date).days
            if 8 <= age <= 30:
                # Determine ISO week
                iso_year, iso_week, _ = log_date.isocalendar()
                weekly_name = f"weekly-{iso_year}-W{iso_week:02d}.md"
                weekly_path = daily_dir / weekly_name

                content = path.read_text(encoding="utf-8")
                with weekly_path.open("a", encoding="utf-8") as f:
                    f.write(f"\n## {path.stem}\n{content}\n")

                path.unlink()
                logger.debug("Compressed %s into %s", path.stem, weekly_name)

    def _archive_old_logs(self) -> None:
        """Move logs older than 30 days to daily/archive/."""
        daily_dir = self._remi.memory.root / "daily"
        archive_dir = daily_dir / "archive"
        now = datetime.now()

        for path in sorted(daily_dir.glob("*.md")):
            if path.stem.startswith("weekly-"):
                try:
                    # Parse weekly-YYYY-WNN
                    parts = path.stem.split("-")
                    year = int(parts[1])
                    week = int(parts[2][1:])
                    week_date = datetime.strptime(f"{year}-W{week:02d}-1", "%Y-W%W-%w")
                    if (now - week_date).days > 30:
                        archive_dir.mkdir(exist_ok=True)
                        path.rename(archive_dir / path.name)
                except (ValueError, IndexError):
                    continue
            else:
                try:
                    log_date = datetime.strptime(path.stem, "%Y-%m-%d")
                    if (now - log_date).days > 30:
                        archive_dir.mkdir(exist_ok=True)
                        path.rename(archive_dir / path.name)
                except ValueError:
                    continue

    async def _cleanup(self) -> None:
        """Remove old daily notes and version files."""
        removed_daily = self._remi.memory.cleanup_old_dailies(keep_days=30)
        removed_versions = self._remi.memory.cleanup_old_versions(keep=50)
        if removed_daily or removed_versions:
            logger.info(
                "Cleanup: removed %d old dailies, %d old versions",
                removed_daily,
                removed_versions,
            )

File: src/scheduler/jobs.ts
/**
 * Scheduler for periodic tasks using pure async patterns.
 *
 * Jobs:
 * - Heartbeat: check provider health
 * - Memory compaction: archive daily notes → long-term memory + entity extraction
 * - Cleanup: remove old dailies and version files
 * - Scheduled skills: config-driven skill execution and push
 */

import type { RemiConfig, ScheduledSkillConfig } from "../config.js";
import { loadConfig } from "../config.js";
import type { Remi } from "../core.js";
import type { Connector } from "../connectors/base.js";
import { createLogger } from "../logger.js";
import {
  existsSync,
  readFileSync,
  writeFileSync,
  mkdirSync,
  readdirSync,
  unlinkSync,
  renameSync,
  appendFileSync,
} from "node:fs";
import { join, dirname } from "node:path";
import { homedir } from "node:os";

const log = createLogger("scheduler");

/** Format a Date as YYYY-MM-DD in local timezone (not UTC). */
function localDateStr(d: Date): string {
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")}`;
}

function parseCronHour(cronExpr: string): number {
  const parts = cronExpr.split(" ");
  if (parts.length >= 2) {
    const hour = parseInt(parts[1], 10);
    if (!isNaN(hour)) return hour;
  }
  return 3; // default: 3 AM
}

export class Scheduler {
  private _remi: Remi;
  private _config: RemiConfig;
  private _compactHour: number;
  private _heartbeatInterval: number;

  constructor(remi: Remi, config: RemiConfig) {
    this._remi = remi;
    this._config = config;
    this._compactHour = parseCronHour(config.scheduler.memoryCompactCron);
    this._heartbeatInterval = config.scheduler.heartbeatInterval;
  }

  async start(shutdownSignal: AbortSignal): Promise<void> {
    const skillNames = this._config.scheduledSkills
      .filter((s) => s.enabled)
      .map((s) => s.name);
    log.info(
      `Scheduler started (heartbeat=${this._heartbeatInterval}s, compact@${String(this._compactHour).padStart(2, "0")}:00` +
        (skillNames.length > 0
          ? `, scheduled_skills: [${skillNames.join(", ")}]`
          : "") +
        ")",
    );

    let lastCompactDate: string | null = null;
    // Per-skill tracking: name → { lastGenDate, lastPushDate }
    const skillTracker = new Map<string, { lastGenDate: string | null; lastPushDate: string | null }>();

    while (!shutdownSignal.aborted) {
      // Wait for heartbeat interval or shutdown
      await new Promise<void>((resolve) => {
        const timer = setTimeout(resolve, this._heartbeatInterval * 1000);
        shutdownSignal.addEventListener(
          "abort",
          () => {
            clearTimeout(timer);
            resolve();
          },
          { once: true },
        );
      });

      if (shutdownSignal.aborted) break;

      // Heartbeat
      await this._heartbeat();

      // Hot-reload configs from remi.toml
      this._reloadScheduledSkillsConfig();

      const now = new Date();
      const today = localDateStr(now);

      // Daily compaction (run once per day at configured hour)
      if (now.getHours() === this._compactHour && lastCompactDate !== today) {
        await this._compactMemory();
        await this._cleanup();
        lastCompactDate = today;
      }

      // Scheduled skills — generic skill execution loop
      for (const skill of this._config.scheduledSkills) {
        if (!skill.enabled || !skill.name) continue;

        if (!skillTracker.has(skill.name)) {
          skillTracker.set(skill.name, { lastGenDate: null, lastPushDate: null });
        }
        const tracker = skillTracker.get(skill.name)!;

        // Generate
        if (
          now.getHours() === skill.generateHour &&
          tracker.lastGenDate !== today
        ) {
          await this._generateSkillReport(skill, today);
          tracker.lastGenDate = today;
        }

        // Push
        if (
          now.getHours() === skill.pushHour &&
          now.getMinutes() >= skill.pushMinute &&
          tracker.lastPushDate !== today
        ) {
          await this._pushSkillReport(skill, today);
          tracker.lastPushDate = today;
        }
      }
    }

    log.info("Scheduler stopped.");
  }

  private async _heartbeat(): Promise<void> {
    for (const [name, provider] of this._remi._providers) {
      try {
        const healthy = await provider.healthCheck();
        if (!healthy) {
          log.warn(`Provider ${name} health check failed`);
        }
      } catch (e) {
        log.error(`Provider ${name} health check error:`, e);
      }
    }

    // Auth token refresh check
    if (this._remi.authStore) {
      try {
        await this._remi.authStore.checkAndRefreshAll();
      } catch (e) {
        log.error("Auth token refresh check error:", e);
      }
    }
  }

  private async _compactMemory(): Promise<void> {
    const yesterday = localDateStr(new Date(Date.now() - 86400000));
    const daily = this._remi.memory.readDaily(yesterday);

    if (!daily || daily.trim().length < 50) {
      return;
    }

    log.info(`Compacting daily notes for ${yesterday}`);

    try {
      const provider = this._remi._getProvider();
      const prompt =
        `Below are my daily notes from ${yesterday}. ` +
        "Extract any important facts, decisions, or preferences that should be " +
        "remembered long-term. Format as bullet points. " +
        "If nothing is worth remembering long-term, respond with 'SKIP'.\n\n" +
        `Also identify any people, organizations, or decisions mentioned. ` +
        `For each, output a line: ENTITY: name (type) - observation\n\n` +
        daily;

      const response = await provider.send(prompt);
      const text = response.text.trim();

      if (text.toUpperCase() !== "SKIP") {
        // Extract entity observations
        for (const line of text.split("\n")) {
          if (line.startsWith("ENTITY:")) {
            this._processEntityLine(line);
          }
        }

        // Filter out ENTITY lines for the general summary
        const summaryLines = text
          .split("\n")
          .filter((line) => !line.startsWith("ENTITY:"));
        const summaryText = summaryLines.join("\n").trim();

        if (summaryText) {
          this._remi.memory.appendMemory(`\n## From ${yesterday}\n\n${summaryText}`);
          log.info(`Appended compacted memory from ${yesterday}`);
        }

        // Update rolling summary
        this._updateRollingSummary(yesterday, summaryText);
      }

      // Compress old daily logs into weekly summaries
      this._compressWeeklyLogs();

      // Archive very old logs
      this._archiveOldLogs();
    } catch (e) {
      log.error("Memory compaction failed:", e);
    }
  }

  private _processEntityLine(line: string): void {
    const match = line.match(/ENTITY:\s*(.+?)\s*\((\w+)\)\s*-\s*(.+)/);
    if (!match) return;
    const [, name, etype, observation] = match;
    try {
      const entityPath = this._remi.memory._findEntityByName(name);
      if (entityPath) {
        this._remi.memory.appendObservation(name, observation);
      } else {
        this._remi.memory.createEntity(name, etype, observation, "agent-inferred");
      }
    } catch (e) {
      log.warn(`Failed to process entity ${name}:`, e);
    }
  }

  private _updateRollingSummary(dateStr: string, summary: string): void {
    const summaryFile = join(this._remi.memory.root, ".conversation_summary.md");
    try {
      let existing = "";
      if (existsSync(summaryFile)) {
        existing = readFileSync(summaryFile, "utf-8");
      }
      const entry = `\n## ${dateStr}\n${summary}\n`;
      writeFileSync(summaryFile, existing + entry, "utf-8");
    } catch (e) {
      log.warn("Failed to update rolling summary:", e);
    }
  }

  private _compressWeeklyLogs(): void {
    const dailyDir = join(this._remi.memory.root, "daily");
    if (!existsSync(dailyDir)) return;

    const now = Date.now();

    for (const file of readdirSync(dailyDir).sort()) {
      if (!file.endsWith(".md") || file.startsWith("weekly-")) continue;
      const stem = file.replace(".md", "");
      const logDate = Date.parse(stem);
      if (isNaN(logDate)) continue;

      const ageDays = (now - logDate) / (24 * 60 * 60 * 1000);
      if (ageDays >= 8 && ageDays <= 30) {
        // Determine ISO week
        const d = new Date(logDate);
        const dayOfYear = Math.floor(
          (d.getTime() - new Date(d.getFullYear(), 0, 0).getTime()) / 86400000,
        );
        const weekNum = Math.ceil(
          (dayOfYear + new Date(d.getFullYear(), 0, 1).getDay()) / 7,
        );
        const weeklyName = `weekly-${d.getFullYear()}-W${String(weekNum).padStart(2, "0")}.md`;
        const weeklyPath = join(dailyDir, weeklyName);

        const content = readFileSync(join(dailyDir, file), "utf-8");
        appendFileSync(weeklyPath, `\n## ${stem}\n${content}\n`, "utf-8");

        unlinkSync(join(dailyDir, file));
      }
    }
  }

  private _archiveOldLogs(): void {
    const dailyDir = join(this._remi.memory.root, "daily");
    if (!existsSync(dailyDir)) return;

    const archiveDir = join(dailyDir, "archive");
    const now = Date.now();

    for (const file of readdirSync(dailyDir).sort()) {
      if (!file.endsWith(".md")) continue;
      const fullPath = join(dailyDir, file);

      if (file.startsWith("weekly-")) {
        try {
          const parts = file.replace(".md", "").split("-");
          const year = parseInt(parts[1], 10);
          const week = parseInt(parts[2].slice(1), 10);
          // Approximate week date
          const weekDate = new Date(year, 0, 1 + (week - 1) * 7);
          if ((now - weekDate.getTime()) / (24 * 60 * 60 * 1000) > 30) {
            if (!existsSync(archiveDir)) mkdirSync(archiveDir, { recursive: true });
            renameSync(fullPath, join(archiveDir, file));
          }
        } catch {
          continue;
        }
      } else {
        const logDate = Date.parse(file.replace(".md", ""));
        if (!isNaN(logDate) && (now - logDate) / (24 * 60 * 60 * 1000) > 30) {
          if (!existsSync(archiveDir)) mkdirSync(archiveDir, { recursive: true });
          renameSync(fullPath, join(archiveDir, file));
        }
      }
    }
  }

  private async _cleanup(): Promise<void> {
    const removedDaily = this._remi.memory.cleanupOldDailies(30);
    const removedVersions = this._remi.memory.cleanupOldVersions(50);
    if (removedDaily || removedVersions) {
      log.info(
        `Cleanup: removed ${removedDaily} old dailies, ${removedVersions} old versions`,
      );
    }
  }

  // ── Scheduled Skills (generic config-driven) ───────────────

  /**
   * Load a skill's SKILL.md content, strip YAML frontmatter,
   * and substitute date placeholders.
   */
  private _loadSkillPrompt(skillName: string, dateStr: string): string | null {
    const skillPath = join(
      homedir(), ".remi", ".claude", "skills", skillName, "SKILL.md",
    );

    if (!existsSync(skillPath)) {
      log.error(`Skill file not found: ${skillPath}`);
      return null;
    }

    let content = readFileSync(skillPath, "utf-8");

    // Strip YAML frontmatter (--- ... ---)
    const frontmatterMatch = content.match(/^---\n[\s\S]*?\n---\n/);
    if (frontmatterMatch) {
      content = content.slice(frontmatterMatch[0].length);
    }

    // Substitute date placeholders
    content = content.replace(/YYYY-MM-DD/g, dateStr);

    return content.trim();
  }

  private _reloadScheduledSkillsConfig(): void {
    try {
      const newConfig = loadConfig();
      const oldSkills = this._config.scheduledSkills;
      const newSkills = newConfig.scheduledSkills;

      if (JSON.stringify(oldSkills) !== JSON.stringify(newSkills)) {
        this._config.scheduledSkills = newSkills;
        this._remi.config.scheduledSkills = newSkills;
        const names = newSkills.filter((s) => s.enabled).map((s) => s.name);
        log.info(`Scheduled skills hot-reloaded: [${names.join(", ")}]`);
      }
    } catch (e) {
      log.warn("Failed to reload scheduled skills config:", e);
    }
  }

  private async _generateSkillReport(skill: ScheduledSkillConfig, today: string): Promise<void> {
    log.info(`Generating scheduled skill report: ${skill.name} for ${today}`);

    try {
      const prompt = this._loadSkillPrompt(skill.name, today);
      if (!prompt) return;

      const provider = this._remi._getProvider();
      const response = await provider.send(prompt);
      const text = response.text.trim();

      if (!text || text.startsWith("[Provider error") || text.startsWith("[Provider timeout")) {
        log.error(`Skill report generation failed (${skill.name}): ${text.slice(0, 100)}`);
        return;
      }

      if (!existsSync(skill.outputDir)) {
        mkdirSync(skill.outputDir, { recursive: true });
      }

      const reportPath = join(skill.outputDir, `${today}.md`);
      writeFileSync(reportPath, text, "utf-8");
      log.info(`Skill report saved: ${skill.name} → ${reportPath}`);
    } catch (e) {
      log.error(`Skill report generation error (${skill.name}):`, e);
    }
  }

  private async _pushSkillReport(skill: ScheduledSkillConfig, today: string): Promise<void> {
    const reportPath = join(skill.outputDir, `${today}.md`);
    if (!existsSync(reportPath)) {
      log.warn(`No report found for skill ${skill.name} on ${today}, skipping push`);
      return;
    }

    const content = readFileSync(reportPath, "utf-8");
    if (!content.trim()) {
      log.warn(`Report for skill ${skill.name} on ${today} is empty, skipping push`);
      return;
    }

    const connectors = this._remi["_connectors"] as Connector[];
    const connector = connectors.find((c) => c.name === skill.connectorName);

    if (!connector) {
      log.error(
        `Skill push (${skill.name}): connector "${skill.connectorName}" not found`,
      );
      return;
    }

    let pushContent = content;

    // Truncate if too long
    if (pushContent.length > skill.maxPushLength) {
      // Try to cut at a section boundary (## heading)
      const truncated = pushContent.slice(0, skill.maxPushLength);
      const lastSection = truncated.lastIndexOf("\n## ");
      const cutPoint = lastSection > skill.maxPushLength * 0.5 ? lastSection : skill.maxPushLength;
      pushContent = pushContent.slice(0, cutPoint).trim() + "\n\n> 回复「完整报告」查看完整内容";
    }

    for (const target of skill.pushTargets) {
      try {
        await connector.reply(target, { text: pushContent });
        log.info(`Skill report pushed: ${skill.name} → ${target}`);
      } catch (e) {
        log.error(`Failed to push skill report (${skill.name}) to ${target}:`, e);
      }
    }
  }
}
